plugins/polio/js/src/domains/VaccineModule/SupplyChain/Details/VAR/VaccineArrivalReport.tsx

```diff
@@ -80,12 +73,124 @@ export const VaccineArrivalReport: FunctionComponent<Props> = ({
+                      )
+                    : 0;
+                setFieldValue(`arrival_reports[${index}].doses_shipped`, value);
+                setFieldValue(
+                    `arrival_reports[${index}].vials_shipped`,
+                    vialsShipped,
+                );
+            }
+        },
+        [doses_per_vial, index, setFieldValue],
+    );
+    const handleDosesReceivedUpdate = useCallback(
+        (value: number) => {
+            if (dosesReceivedRef.current) {
+                const vialsReceived = doses_per_vial
+                    ? Math.ceil(
+                          ((value as Optional<number>) ?? 0) / doses_per_vial,
+                      )
+                    : 0;
+                setFieldValue(
+                    `arrival_reports[${index}].doses_received`,
+                    value,
+                );
+                setFieldValue(
+                    `arrival_reports[${index}].vials_received`,
+                    vialsReceived,
+                );
+            }
+        },
+        [doses_per_vial, index, setFieldValue],
+    );
+
+    const handleVialsShippededUpdate = useCallback(
+        (value: number) => {
+            if (vialsShippedRef.current) {
+                const dosesShipped = value * (doses_per_vial ?? 0);
+                setFieldValue(`arrival_reports[${index}].vials_shipped`, value);
+                setFieldValue(
+                    `arrival_reports[${index}].doses_shipped`,
+                    dosesShipped,
+                );
+            }
         },
-        [index, setFieldTouched, setFieldValue, values.pre_alerts],
+        [doses_per_vial, index, setFieldValue],
     );
+    const handleVialsReceivedUpdate = useCallback(
+        (value: number) => {
+            if (vialsReceivedRef.current) {
+                const dosesReceived = value * (doses_per_vial ?? 0);
+                setFieldValue(
+                    `arrival_reports[${index}].vials_received`,
+                    value,
+                );
+                setFieldValue(
+                    `arrival_reports[${index}].doses_received`,
+                    dosesReceived,
+                );
+            }
+        },
+        [doses_per_vial, index, setFieldValue],
+    );
+    const onDosesShippedFocused = () => {
+        dosesShippedRef.current = true;
+    };
+    const onDosesShippedBlur = () => {
+        dosesShippedRef.current = false;
+    };
+    const onDosesReceivedFocused = () => {
+        dosesReceivedRef.current = true;
+    };
+    const onDosesReceivedBlur = () => {
+        dosesReceivedRef.current = false;
+    };
+    const onVialsShippedFocused = () => {
+        vialsShippedRef.current = true;
+    };
+    const onVialsShippedBlur = () => {
+        vialsShippedRef.current = false;
+    };
+    const onVialsReceivedFocused = () => {
+        vialsReceivedRef.current = true;
+    };
+    const onVialsReceivedBlur = () => {
+        vialsReceivedRef.current = false;
+    };
```

Comment:
It should be possible to improve this ...

Reply:
probably, but since the component is super complicated already, so I thought better to keep it straightforward, even if it creates some repetition